[Diagnostics] Add -no-warning-as-error to except a specific warning from being treated as an error

DmT021 opened the pull request:
<!-- What's in this pull request? -->
This commit adds a new compiler option `-no-warning-as-error` which allows users to specify exceptions from the `-warnings-as-errors` option. 
The format of the option is `-no-warning-as-error <diag_id>`, where <diag_id> is one of the diagnostic IDs listed in `include/swift/AST/Diagnostics*.def`. 
The option can be added multiple times.

<!--
If this pull request resolves any GitHub issues, link them.
For information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

DmT021 committed:
[Diagnostics] Add -[no-]warning-as-error flags for precise control over warning behavior

This commit adds new compiler options -no-warning-as-error/-warning-as-error which allows users to specify behavior for exact warnings and warning groups.

DmT021 commented:
@artemcm Can you take a look please?

DmT021 commented:
@hborla Can you take a look please?

artemcm commented on include/swift/Basic/DiagnosticOptions.h:
```diff
@@ -61,6 +62,9 @@ class DiagnosticOptions {
   /// Treat all warnings as errors
   bool WarningsAsErrors = false;
 
+  /// Don't treat these warnings as errors when WarningsAsErrors=true
```
Please document that these are diagnosticID strings here.

artemcm reviewed (approved):
Sorry for the delay @DmT021.
This change looks good, and I think it's good to keep this in `HelpHidden`. 

Please also make a matching change in:
https://github.com/apple/swift-driver?tab=readme-ov-file#rebuilding-optionsswift

artemcm commented:
It's worth noting that there are no guarantees that diagnostic IDs are stable API between different compiler versions, they have not historically been treated as such.

DmT021 commented:
> It's worth noting that there are no guarantees that diagnostic IDs are stable API between different compiler versions, they have not historically been treated as such.

Yeah, but when used with `-debug-diagnostic-names` it's pretty easy to see what the diagnosticID is of a particular warning. Should I include this note in the help text as well?

artemcm commented:
> > It's worth noting that there are no guarantees that diagnostic IDs are stable API between different compiler versions, they have not historically been treated as such.
> 
> Yeah, but when used with `-debug-diagnostic-names` it's pretty easy to see what the diagnosticID is for a particular warning. Should I include this note in the help text as well?

Yes, please.

artemcm reviewed (commented):
After consulting with some more folks, I think the use of diagnostic IDs is a bigger concern. 

Though the idea behind this change is worthwhile.
One possible approach to get the desired result is to implement something akin to Clang's diagnostic groups. Then warning diagnostics can be grouped into stable-named groups which can then be treated as exceptions to warning-as-error, for example.

DougGregor commented:
> It's worth noting that there are no guarantees that diagnostic IDs are stable API between different compiler versions, they have not historically been treated as such.

This is a really important point. The addition of this feature would hamper further evolution of the compiler, because code that relies on `-no-warning-as-error <diag_id>` to compile will stop working if the diagnostic ID changes. We can't just promise not to make changes, because it's very common for us to improve diagnostics by adding more special-case versions of existing warnings, which will necessitate new diagnostic IDs.

I support the general idea of more specific control over warnings, but not based on the diagnostic IDs. Clang's approach of having documented "warning groups" with stable IDs, where each warning is sorted into an appropriate warning group, would be a reasonable model to follow. It's definitely more upfront work to do this... along with building the infrastructure for warning groups, we'd also need to sort the existing ~350 warnings into groups. But if we're going to do this, it's worth doing it well so we can provide flexibility to developers while still being able to evolve the compiler.

DmT021 commented:
Using groups will result in a loss of precision when defining exclusion rules, which isn't desirable either. Of course, I agree that control via diagnostic IDs isn't very rigid, and group-based rules are superior. But this is only important if the user wants to distribute the library. If they are building an end binary, it is quite acceptable.
This change addresses existing situations where a newly diagnosed warning by a new compiler led to a failed build because of the `-warnings-as-errors` rule.
What if we provide a generalized rule syntax that allows the user to specify a condition (either an ID or a group) to check a warning against? Something like
```
-no-warning-as-error ignored_import                     (group)
-no-warning-as-error id:sema_import_current_module      (id)
```
Using id-based rules is still discouraged for source-distributed libraries, but is still an option for everyone else.
And then we can categorize warnings into groups gradually.

DougGregor commented:
> Using groups will result in a loss of precision when defining exclusion rules, which isn't desirable either

It's necessary to allow us to improve diagnostics in the future, but providing more-specific diagnostics in certain cases.

> But this is only important if the user wants to distribute the library. If they are building an end binary, it is quite acceptable.

That's not true. As I described in my original reply, we must preserve the ability to create more specific versions of a given warning that provide a better diagnosis for the user. Adding this feature based on diagnostic IDs makes that kind of change a source-breaking change.

DougGregor commented:
@DmT021 there are a bunch of design and usability questions here, and it's grown beyond what's reasonable to figure out in the pull request. I think we'd benefit from taking the discussion over to forums.swift.org where we can get more feedback from others

DmT021 commented:
https://forums.swift.org/t/warnings-as-errors-exceptions/72925

DmT021 commented:
@artemcm @DougGregor This is still work-in-progress but I want to check if I'm moving in the right direction. Here's what I have so far:
- [DiagnosticGroups.def](https://github.com/swiftlang/swift/pull/74466/files#diff-ad677bf180808d16a1726a4255dbe362641c925f2f256fe24cb7f9c0d691578b) file defines diagnostic groups. I took clang's approach and made groups nested: a group can have subgroups.
- To have something to begin with I declared several groups. In particular there are the `deprecated` group and `availability_deprecated` which is nested to `deprecated`. The idea here is to have all deprecation warnings nested in the `deprecated` group so they can be excluded together.
- It compiles as constexpr (almost, unfortunately llvm::ArrayRef can't be constexpr).
- `WARNING` and `ERROR` diagnostics now have `GROUPED_` variants.
- New compiler options `-[no-]-warning-as-error <group>`. `<group>` can be either a real group or a string `_id:<diag_id>`. And a new option group `warning_treating_Group` that aggregates them.
- `-[no-]-warning-as-error` and `-[no-]-warnings-as-errors` apply in order they appear in the command line. If there're more than one rule affecting the behavior of the same DiagID - the last rule wins.
- The Diagnostic Groups graph consistency expressed as `static_assert`s and not dedicated tests.

What's left to do:
- It seems reasonable to include the "warning suppression" rules in this mechanism. But I decided to avoid doing this until we agree on it.
- The diagnostic groups graph should be checked for cycles.
- There's a difficulty with testing that a particular warning is actually in the expected group(s). 
  For example there's a bunch of test for the `@available(*, deprecated)` annotation. We need to check that all of them are in the `availability_deprecated` and `deprecated` groups. We can invoke the compiler with all combinations of `-[no-]-warning-as-error` and `-[no-]-warnings-as-errors` flags for each test, but this will result in combinatorial explosion. 
  So I came up with an alternative: we can introduce a new flag `-pring-diagnostic-groups` which will add a suffix with `[group:<group_name>]` to all the warnings produced. We only need to print the group that directly includes this particular `DiagID` and not the whole hierarchy of groups. Then we test that warnings are suffixed with the expected groups.
  (We should also test that groups are nested correctly, but that's already done with `static_assert`s.)
  Is it ok?

DmT021 commented:
@artemcm @DougGregor ping?

DougGregor commented on include/swift/AST/DiagnosticsCommon.def:
```diff
@@ -47,7 +47,7 @@ ERROR(error_no_group_info,none,
      "did you mean to use a 'do' statement?", ())
 
-WARNING(error_in_future_swift_version,none,
+GROUPED_WARNING(error_in_future_swift_version,error_in_future_swift_version,none,
```
This specific warning is a bit tricky, because everything that goes through it started its life as an error.

DmT021 replied:
Yep, it confuses me a little. It's not technically wrong - if a warning was upgraded to error because we enabled `-swift 6` we want it to be an actual error and fail the build. But it would be nice if there was some cleaner solution for these warnings.

DougGregor replied:
The unfortunate thing about this warning group is that it's going to mix unrelated diagnostics together into a single warning group, because we use it via the `warnUntilSwiftVersion(N)` on ERRORs of various flavors. The only other idea I have here is to (eventually) require ERRORs treated this way to be part of a warning group. For example, a "concurrency" warning group could encapsulate all of the concurrency diagnostics (which are ERRORs) when they get downgraded. But we'd have to disallow them from being downgraded in Swift 6 somehow, so there's a bit of work to get to that point.

DmT021 replied:
It was my initial idea too, which is why both `GROUPED_WARNING` and `GROUPED_ERROR` exist. The issue is that, theoretically, a `Diagnostic` can be nested inside another `Diagnostic` at any place within the `Args` vector. And this tree can be of any depth. If we want to check nested diagnostics we have to traverse all the tree.
And then if at least one diagnostic with an excluded `groupID` is found - apply the behavior(error/warning/suppress?) to the root `Diagnostic`.
I wasn't sure if it's correct.

DougGregor commented on lib/AST/DiagnosticGroups.cpp:
```diff
@@ -0,0 +1,236 @@
+  return false;
+}
+
+static_assert(!hasCycle(), "Diagnostic groups graph has a cycle!");
```
This is quite some clever metaprogramming.

DougGregor commented on test/diagnostics/warnings_as_errors_rules.swift:
```diff
@@ -0,0 +1,65 @@
+// RUN: not %target-swift-frontend -typecheck -diagnostic-style llvm -warnings-as-errors %s 2>&1 | %FileCheck %s --check-prefix=CHECK-WAE-ALL
```
Why is this forcing the LLVM diagnostic style?

DmT021 replied:
With the swift style the compiler produces a message like this:
```
$ ./swift-frontend -typecheck -diagnostic-style swift -warnings-as-errors  ../../../../swift/test/diagnostics/warnings_as_errors_rules.swift
../../../../swift/test/diagnostics/warnings_as_errors_rules.swift:42:1: error: 'foo()' is deprecated
40 | // CHECK-WAE-GROUP-NWAE-DIAGID: warning: 'foo()' is deprecated [group:availability_deprecated]
41 | // CHECK-WAE-GROUP-NWAE-DIAGID-NOT: error: 'foo()' is deprecated [group:availability_deprecated]
42 | foo()
   | `- error: 'foo()' is deprecated
43 |
44 |
```
As you can see there are two lines more above and below the line 42. FileCheck sees it in its input and when it performs a `-NOT` check these comments actually produce a match (which fails the testcase). 
Ideally I'd prefer if FileCheck ignored lines containing `.*CHECK-.*`, but I don't see a flag or option for this.

DougGregor replied:
Ah, I understand. In our call through to the swift-syntax diagnostic mechanism, we're hard-coding the context size to 2 lines:

```
  swift_ASTGen_renderQueuedDiagnostics(queuedDiagnostics, /*contextSize=*/2,
                                       forceColors ? 1 : 0,
                                       &bridgedRenderedString);
```

... we should have an option here to allow setting the context size (e.g., to 0), which would make FileCheck tests like this more feasible.

DougGregor replied:
To be clear, I don't mean "you have to add this other command-line argument yourself." I'm fine with your PR using `-diagnostic-style llvm` until we get the other command-line argument. I suspect that's the reason for some of the other existing uses of `-diagnostic-style llvm`

DougGregor reviewed (commented):
I'll (separately) comment on the design aspects, but the implementation here looks good and I only have a few small comments.

DougGregor commented:
@swift-ci please smoke test

DmT021 commented on include/swift/AST/DiagnosticGroups.h:
```diff
@@ -0,0 +1,61 @@
+
+struct DiagGroupInfo {
+  DiagGroupID id;
+  std::string_view name;
```
I struggle to see when to use `llvm::StringRef` and when `std::string_view`. It seems both of them are just a pair size+pointer and both have "not owning" semantics. Should I just stick to llvm types when available?

DougGregor replied:
Let's see how CI goes. I don't have a sense of the standard library support we can depend on, so I tend to stick with the LLVM types.

DmT021 commented:
MSVC failed on __VA_OPT__. I might need some time to find a windows machine and come up with an alternative.

DmT021 commented:
The changes:
- Removed `__VA_OPT__`
I couldn't find an acceptable way to iterate over `__VA_ARGS__` without `__VA_OPT__`. Most solutions fail to compile on MSVC. So I changed the syntax for declaring groups by splitting it into two macros: `GROUP` and `GROUP_LINK`. Once we switch to C++20, we can bring `__VA_OPT__` back.
- Removed support for `_id:`
- Used `llvm::function_ref`
- Added docs and examples for each diagnostic group. 
- Removed support for `error_in_future_swift_version`. I'll work on it separately.
- Removed `#pragma clang` and used [`[[maybe_unused]]`](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused) attribute instead.

@DougGregor Can you take a look and rerun the tests, please?

tshortli commented:
@swift-ci please test

DmT021 commented:
The fix for broken Linux builds has been [merged](https://github.com/swiftlang/swift-integration-tests/pull/143). Can someone rerun CI?

tshortli commented:
@swift-ci please test Linux

DougGregor commented on lib/Driver/ToolChains.cpp:
```diff
@@ -311,8 +311,7 @@ void ToolChain::addCommonFrontendArgs(const OutputInfo &OI,
   inputArgs.AddLastArg(arguments, options::OPT_profile_coverage_mapping);
-  inputArgs.AddAllArgs(arguments, options::OPT_warnings_as_errors,
-                       options::OPT_no_warnings_as_errors);
+  inputArgs.AddAllArgs(arguments, options::OPT_warning_treating_Group);
```
We'll (separately) need to do this in the new swift-driver

DmT021 replied:
Ok, here's a PR https://github.com/swiftlang/swift-driver/pull/1678

DougGregor reviewed (commented):
This looks really good; I have no further comments about the implementation that aren't "ooh, and then we can do <next nice thing>!"

DmT021 commented:
> This looks really good; I have no further comments about the implementation that aren't "ooh, and then we can do !"

Should I ping other folks to review this as well? Or I can just squash rebase and you'll merge it?

DougGregor commented:
> > This looks really good; I have no further comments about the implementation that aren't "ooh, and then we can do !"
> 
> @DougGregor Should I ping other folks to review this as well? Or I can just squash rebase and you'll merge it?

To cover all of the bases, I've brought up this feature with the Language Steering Group to determine whether we should review it through the evolution process. I suspect we don't, because it's primarily a compiler feature that's not exposed through the language, but it's enough of a change in direction that I want to be sure to involve them.

mtzaquia commented:
> > > This looks really good; I have no further comments about the implementation that aren't "ooh, and then we can do !"
> > 
> > 
> > @DougGregor Should I ping other folks to review this as well? Or I can just squash rebase and you'll merge it?
> 
> To cover all of the bases, I've brought up this feature with the Language Steering Group to determine whether we should review it through the evolution process. I suspect we don't, because it's primarily a compiler feature that's not exposed through the language, but it's enough of a change in direction that I want to be sure to involve them.

Hey @DougGregor, do you have any updates here? It would be great to have this merged as soon as you come to an agreement. Also, would you be able to provide a rough estimate on when these changes would land on a RELEASE build of Swift?

DmT021 commented:
@mtzaquia This feature has recently expanded beyond what is implemented in this PR and will go through a review.

DougGregor commented:
The proposal has been accepted!

DmT021 commented:
I'm aligning the implementation here with what the proposal says.

DmT021 committed:
fixup! [Diagnostics] Add -[no-]warning-as-error flags for precise control over warning behavior

DmT021 commented:
I fixed the flag names, added a check to the old driver to prevent the new flags from being used with `-suppress-warnings`, and updated the tests. I'll implement `-print-diagnostic-groups` in the next PR, but other than that, everything here is now in line with the proposal.

Someone, please rerun the tests 🙏

DougGregor commented:
@swift-ci please test

DmT021 commented:
The errors on macOS are rather strange. It doesn't look like it's because of this PR: `'Mirror' is unavailable`, `broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>`, `failed to build module 'Darwin'; this SDK is not supported by the compiler`. And I see other recently failed builds [here](https://ci.swift.org/job/swift-PR-macos/) with the same errors.
<details>
```
FAILED: stdlib/private/SwiftPrivate/FREESTANDING/arm64/SwiftPrivate.o /Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/build/buildbot_incremental/minimalstdlib-macosx-x86_64/stdlib/private/SwiftPrivate/FREESTANDING/arm64/SwiftPrivate.o 
...
/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:29:34: error: 'Mirror' is unavailable
 27 | }
 28 | extension Darwin.DarwinBoolean : Swift.CustomReflectable {
 29 |   public var customMirror: Swift.Mirror {
    |                                  `- error: 'Mirror' is unavailable
 30 |     get
 31 |   }

Swift.Mirror:2:15: note: 'Mirror' has been explicitly marked unavailable here
 1 | @available(*, unavailable)
 2 | public struct Mirror {
   |               `- note: 'Mirror' has been explicitly marked unavailable here
 3 |     public enum AncestorRepresentation {
 4 |         case generated

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:28:40: error: 'CustomReflectable' is unavailable
 26 |   public typealias BooleanLiteralType = Swift.Bool
 27 | }
 28 | extension Darwin.DarwinBoolean : Swift.CustomReflectable {
    |                                        `- error: 'CustomReflectable' is unavailable
 29 |   public var customMirror: Swift.Mirror {
 30 |     get

Swift.CustomReflectable:2:17: note: 'CustomReflectable' has been explicitly marked unavailable here
1 | @available(*, unavailable)
2 | public protocol CustomReflectable {
  |                 `- note: 'CustomReflectable' has been explicitly marked unavailable here
3 |     var customMirror: Mirror { get }
4 | }

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:767:36: error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
765 | @_transparent public func lgamma(_ x: Swift.Float) -> (Swift.Float, Swift.Int) {
766 |   var sign = Int32(0)
767 |   let value = lgammaf_r(CFloat(x), &sign)
    |                                    `- error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
768 |   return (Float(value), Int(sign))
769 | }

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:772:36: error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
770 | @_transparent public func lgamma(_ x: Swift.Double) -> (Swift.Double, Swift.Int) {
771 |   var sign = Int32(0)
772 |   let value = lgamma_r(CDouble(x), &sign)
    |                                    `- error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
773 |   return (Double(value), Int(sign))
774 | }

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:777:43: error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
775 | @_transparent public func remquo(_ x: Swift.Float, _ y: Swift.Float) -> (Swift.Float, Swift.Int) {
776 |   var quo = Int32(0)
777 |   let rem = remquof(CFloat(x), CFloat(y), &quo)
    |                                           `- error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
778 |   return (Float(rem), Int(quo))
779 | }

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:782:44: error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
780 | @_transparent public func remquo(_ x: Swift.Double, _ y: Swift.Double) -> (Swift.Double, Swift.Int) {
781 |   var quo = Int32(0)
782 |   let rem = remquo(CDouble(x), CDouble(y), &quo)
    |                                            `- error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
783 |   return (Double(rem), Int(quo))
784 | }

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:787:21: error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
785 | @available(swift, deprecated: 4.2, message: "use Float(nan: Float.RawSignificand).")
786 | @_transparent public func nan(_ tag: Swift.String) -> Swift.Float {
787 |   return Float(nanf(tag))
    |                     `- error: broken standard library: cannot find intrinsic operations on UnsafeMutablePointer<T>
788 | }
789 | @_transparent public func jn(_ n: Swift.Int, _ x: Swift.Double) -> Swift.Double {

/Users/ec2-user/jenkins/workspace/swift-PR-macos/branch-main/swift/stdlib/private/SwiftPrivate/IO.swift:21:8: error: failed to build module 'Darwin'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 5.9.2 (swiftlang-5.9.2.2.11 clang-1500.1.0.2.2)', while this compiler is 'Apple Swift version 6.1-dev effective-5.10 (LLVM 80d178d8cbb1cee, Swift 67ac5b9e08aa6ea)'). Please select a toolchain which matches the SDK.
 19 | #else
 20 | #if canImport(Darwin)
 21 | import Darwin
    |        `- error: failed to build module 'Darwin'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 5.9.2 (swiftlang-5.9.2.2.11 clang-1500.1.0.2.2)', while this compiler is 'Apple Swift version 6.1-dev effective-5.10 (LLVM 80d178d8cbb1cee, Swift 67ac5b9e08aa6ea)'). Please select a toolchain which matches the SDK.
 22 | #elseif canImport(Glibc)
 23 | import Glibc

/Applications/Xcode-beta.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/lib/swift/Darwin.swiftmodule/arm64e-apple-macos.swiftinterface:29:34: error: 'Mirror' is unavailable
 27 | }
 28 | extension Darwin.DarwinBoolean : Swift.CustomReflectable {
 29 |   public var customMirror: Swift.Mirror {
    |                                  `- error: 'Mirror' is unavailable
 30 |     get
 31 |   }
```
</details>

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented:
The macOS errors you are seeing do seem unrelated to your change.

DougGregor commented:
@swift-ci please smoke test macOS

DougGregor commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -851,8 +852,8 @@ namespace swift {
 
-    /// Emit all warnings as errors
-    bool warningsAsErrors = false;
+    /// Treat these warnings as errors. Indicies here corespond to DiagID enum
```
Typos "indicies" (for indices) and "corespond"

DougGregor commented on include/swift/AST/DiagnosticEngine.h:
```diff
@@ -1105,11 +1119,13 @@ namespace swift {
-      return state.getWarningsAsErrors();
-    }
+    /// Apply rules specifing what warnings should or shouldn't be treated as
+    /// errors. For group rules the string is either a group name defined by
```
There's no "or" to go with the "either here".

DougGregor commented on userdocs/diagnostic_groups/availability_deprecated.md:
````diff
@@ -0,0 +1,30 @@
+## Usage Example
+
+```sh
+swiftc -warning-as-error availability_deprecated file.swift
+swiftc -warnings-as-errors -no-warning-as-error availability_deprecated file.swift
````
These flag names have changed in SE-0443.

DougGregor commented on userdocs/diagnostic_groups/deprecated.md:
````diff
@@ -0,0 +1,12 @@
+## Usage Example
+
+```sh
+swiftc -warning-as-error deprecated file.swift
+swiftc -warnings-as-errors -no-warning-as-error deprecated file.swift
````
Flag names have changed.

DougGregor commented on userdocs/diagnostic_groups/unknown_warning_group.md:
````diff
@@ -0,0 +1,15 @@
+Such warnings are emitted after the behavior for all specified warning groups has been processed, which means their behavior can also be specified. For example:
+
+```sh
+swiftc -warning-as-error unknown_warning_group -warning-as-error non_existing_group file.swift
````
Flag names again

DougGregor reviewed (approved):
All my remaining comments are cosmetic, and the proposal has been accepted. Let's merge this!

AakashSorathiya commented:
I can see not all diagnostic IDs are associated with diagnostic groups (for ex. `nonoptional_compare_to_nil`). So, is there any plan to associate all IDs to groups? Or is there any other option to handle such IDs separately?

AakashSorathiya commented:
> > I can see not all diagnostic IDs are associated with diagnostic groups (for ex. `nonoptional_compare_to_nil`). So, is there any plan to associate all IDs to groups? Or is there any other option to handle such IDs separately?
> 
> We'd like to provide groups for all warnings (and, eventually, all diagnostics). Doing so also means documenting these groups, as described in https://github.com/swiftlang/swift/blob/main/docs/Diagnostics.md#diagnostic-groups

Okay, thank you for replying. I will also try to contribute to this work. 

Also, quick question - does `-Werror` and `-Wwarning` flags work with `-Xfrontend` ?
